[WebGPU] GPUBuffer::getMappedRange will attempt to copy from nullptr if mapping fails

mwyrzykowski opened the pull request:
#### fee8a92ad0b7669ae5d66fa7d0e59a578ef6f327
<pre>
[WebGPU] GPUBuffer::getMappedRange will attempt to copy from nullptr if mapping fails
<a href="https://bugs.webkit.org/show_bug.cgi?id=252603">https://bugs.webkit.org/show_bug.cgi?id=252603</a>
&lt;radar://105692446&gt;

Reviewed by Myles C. Maxfield.

Add some additional validation to improve conformance with CTS
and prevent errors during invalid API usage.

Merged some changes from PR <a href="https://github.com/WebKit/WebKit/pull/714">https://github.com/WebKit/WebKit/pull/714</a>
authored by Myles C. Maxfield.

* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUBufferImpl.cpp:
(PAL::WebGPU::getMappedSize):
(PAL::WebGPU::BufferImpl::getMappedRange):

* Source/WebGPU/WebGPU/Buffer.mm:
(WebGPU::Device::createBuffer):
(WebGPU::Buffer::Buffer):
(WebGPU::Buffer::getMappedRange):

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.cpp:
(WebKit::RemoteBuffer::mapAsync):

* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBufferProxy.cpp:
(WebKit::WebGPU::RemoteBufferProxy::mapAsync):
(WebKit::WebGPU::offsetOrSizeExceedsBounds):
(WebKit::WebGPU::RemoteBufferProxy::getMappedRange):

* LayoutTests/http/tests/webgpu/webgpu/api/operation/buffers/map-expected.txt:
* LayoutTests/platform/mac/TestExpectations:
Enable new test.

Canonical link: <a href="https://commits.webkit.org/261555@main">https://commits.webkit.org/261555@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/12f70d8eec6254f7b1752dd449d52aa806e62a1d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112087 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21218 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/710 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3883 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120746 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116148 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22566 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/12349 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/3952 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117848 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/16767 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99932 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105144 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98723 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/479 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45770 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13633 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/512 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/93384 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14311 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/12349 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19683 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52511 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8054 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16102 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/Buffer.mm:
```diff
@@ -219,8 +228,6 @@ static size_t computeRangeSize(uint64_t size, size_t offset)
     if (!isValid())
         return false;
 
-    // FIXME: The spec says "TODO: check destroyed state?"
```
comment is not needed, that is done by 'isValid'

djg commented on Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUBufferImpl.cpp:
```diff
@@ -45,11 +45,20 @@ BufferImpl::~BufferImpl()
 
+static Size64 getSizeFromBacking(WGPUBuffer backing, std::optional<Size64> size, Size64 offset)
+{
+    if (size.has_value())
+        return size.value();
```
So, if `size` is a value, it doesn't get `getSizeFromBacking`? How about `getMappedSize(m_backing, size, offset)`

djg reviewed (commented):
:shipit:

litherum commented on Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUBufferImpl.cpp:
```diff
@@ -45,11 +45,20 @@ BufferImpl::~BufferImpl()
 
-    auto usedSize = size.value_or(WGPU_WHOLE_MAP_SIZE);
+    auto usedSize = getMappedSize(m_backing, size, offset);
 
```
I feel like we're missing a bunch of validation here...

mwyrzykowski replied:
The validation happens at the WebGPU.framework layer

litherum commented on Source/WebGPU/WebGPU/Buffer.mm:
```diff
@@ -120,6 +120,7 @@ static MTLStorageMode storageMode(bool deviceHasUnifiedMemory, WGPUBufferUsageFl
         return Buffer::createInvalid(*this);
 
     buffer.label = fromAPI(descriptor.label);
+    ASSERT(buffer.contents);
```
storage mode can be private, isn't this wrong?

mwyrzykowski replied:
Oh good point, I’ll remove these assertions

litherum commented on Source/WebGPU/WebGPU/Buffer.mm:
```diff
@@ -135,6 +136,7 @@ static MTLStorageMode storageMode(bool deviceHasUnifiedMemory, WGPUBufferUsageFl
     , m_mappingRange(initialMappingRange)
     , m_device(device)
 {
+    ASSERT(m_buffer.contents);
```
Ditto

litherum reviewed (approved):
How similar is this to https://github.com/WebKit/WebKit/pull/714/files#diff-cae273a4d92911c86170f1292baa2d5c7b2ceab172d93aba68e736a6ce4c6cc9? Can we try to merge that into this somehow?

cdumez commented on Source/WebCore/Modules/WebGPU/GPUBuffer.cpp:
```diff
@@ -42,15 +56,48 @@ void GPUBuffer::setLabel(String&& label)
+        m_mapState = GPUBufferMapState::Pending;
+
+    m_pendingMapPromise = promise;
+    m_backing->mapAsync(convertMapModeFlagsToBacking(mode), offset.value_or(0), size, [promise = WTFMove(promise), strongThis = Ref { *this }] (bool success) mutable {
```
We normally use `protectedThis` in C++ code (and `strongSelf` in ObjC++ code).

mwyrzykowski replied:
Thanks I'll change that

cdumez commented on Source/WebCore/Modules/WebGPU/GPUBufferMapState.idl:
```diff
@@ -0,0 +1,35 @@
+[
+    EnabledBySetting=WebGPU
+]
+enum GPUBufferMapState {
```
We usually but this on the previous line:
`] enum GPUBufferMapState {`

mwyrzykowski replied:
Should I change that as all of the existing WebGPU enums are on separate lines? There are 35 or so which can be found by searching `enum GPU`

cdumez replied:
I'm fine either way but it'd be good to align WebGPU enums with the style we use for the rest of WebKit.

mwyrzykowski replied:
Sounds good, I filed https://bugs.webkit.org/show_bug.cgi?id=253740

and will clean them all up to be consistent in a future patch

cdumez commented on Source/WebGPU/WebGPU/Buffer.mm:
```diff
@@ -337,6 +347,11 @@ static size_t computeRangeSize(uint64_t size, size_t offset)
 
+uint64_t Buffer::size() const
+{
+    return m_emptyBuffer.size() ?: m_size;
```
is `?:` standard C++? I thought it was a GNU extension but maybe thing have changed.

mwyrzykowski replied:
I don't think it is, but we are in Objective-C++ so I think this is ok as it is widely used in Objective-C/C++

I can also avoid using it too if you prefer

cdumez replied:
Oh duh, my bad. Failed to notice this was a mm file :)

cdumez commented on Source/WebGPU/WebGPU/Device.mm:
```diff
@@ -254,6 +254,25 @@ static void captureFrame(id<MTLDevice> captureObject)
+    }
+
+    if (m_uncapturedErrorCallback) {
+        instance().scheduleWork([protectedThis = Ref { *this }, message = WTFMove(message)]() mutable {
```
Hopefully message will be used on the same thread? (otherwise you need an isolatedCopy()).

mwyrzykowski replied:
Correct it is the same thread. I don't think there is an obvious need to do this asynchronously as the operation is already async from the clients perspective, but there is some ambiguity in the specification and a previous discussion with @litherum reached the conclusion it should be async

cdumez commented on Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBufferProxy.cpp:
```diff
@@ -44,31 +44,42 @@ RemoteBufferProxy::~RemoteBufferProxy()
-        callback();
-        return;
-    }
+    auto sendResult = sendWithAsyncReply(Messages::RemoteBuffer::MapAsync(mapModeFlags, offset, size), [callback = WTFMove(callback), mapModeFlags, strongThis = Ref { *this }](auto data) mutable {
```
`protectedThis`

mwyrzykowski committed:
[WebGPU] GPUBuffer::getMappedRange will attempt to copy from nullptr if mapping fails
https://bugs.webkit.org/show_bug.cgi?id=252603
<radar://105692446>

Reviewed by Myles C. Maxfield.

Add some additional validation to improve conformance with CTS
and prevent errors during invalid API usage.

Merged some changes from PR https://github.com/WebKit/WebKit/pull/714
authored by Myles C. Maxfield.

* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUBufferImpl.cpp:
(PAL::WebGPU::getMappedSize):
(PAL::WebGPU::BufferImpl::getMappedRange):

* Source/WebGPU/WebGPU/Buffer.mm:
(WebGPU::Device::createBuffer):
(WebGPU::Buffer::Buffer):
(WebGPU::Buffer::getMappedRange):

* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.cpp:
(WebKit::RemoteBuffer::mapAsync):

* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBufferProxy.cpp:
(WebKit::WebGPU::RemoteBufferProxy::mapAsync):
(WebKit::WebGPU::offsetOrSizeExceedsBounds):
(WebKit::WebGPU::RemoteBufferProxy::getMappedRange):

* LayoutTests/http/tests/webgpu/webgpu/api/operation/buffers/map-expected.txt:
* LayoutTests/platform/mac/TestExpectations:
Enable new test.

Canonical link: https://commits.webkit.org/261555@main
